File: src/main/java/gg/mineral/practice/inventory/menus/SelectModeMenu.java
package gg.mineral.practice.inventory.menus;

import gg.mineral.practice.inventory.ClickCancelled;
import gg.mineral.practice.inventory.PracticeMenu;
import gg.mineral.practice.inventory.SubmitAction;
import gg.mineral.practice.util.items.ItemStacks;
import gg.mineral.practice.util.messages.CC;
import lombok.RequiredArgsConstructor;
import lombok.val;

@ClickCancelled(true)
@RequiredArgsConstructor
public class SelectModeMenu extends PracticeMenu {
	private final SubmitAction action;

	@Override
	public void update() {

		setSlot(2, ItemStacks.SIMPLE_MODE, interaction -> {
			val p = interaction.getProfile();
			p.resetQueueSettings();
			p.openMenu(new SelectExistingKitMenu(new SelectArenaMenu(action), true));
		});

		setSlot(6, ItemStacks.ADVANCED_MODE,
				interaction -> interaction.getProfile().openMenu(new MechanicsMenu(action)));
	}

	@Override
	public String getTitle() {
		return CC.BLUE + "Select Mode";
	}

	@Override
	public boolean shouldUpdate() {
		return true;
	}
}

File: src/main/kotlin/gg/mineral/practice/inventory/menus/SelectModeMenu.kt
package gg.mineral.practice.inventory.menus

import gg.mineral.practice.entity.PlayerStatus
import gg.mineral.practice.inventory.ClickCancelled
import gg.mineral.practice.inventory.Interaction
import gg.mineral.practice.inventory.PracticeMenu
import gg.mineral.practice.inventory.SubmitAction
import gg.mineral.practice.managers.QueuetypeManager.queuetypes
import gg.mineral.practice.util.items.ItemStacks
import gg.mineral.practice.util.messages.CC

@ClickCancelled(true)
class SelectModeMenu(private val action: SubmitAction) : PracticeMenu() {

    override fun update() {
        setSlot(2, ItemStacks.SIMPLE_MODE) { interaction: Interaction ->
            val p = interaction.profile
            p.resetDuelSettings()
            p.openMenu(
                SelectExistingKitMenu(
                    {
                        if (viewer.playerStatus === PlayerStatus.KIT_CREATOR) {
                            viewer.giveKit(it.kit)
                            return@SelectExistingKitMenu
                        }

                        viewer.duelSettings.gametype = it
                        viewer.duelSettings.kit = it.kit

                        val queuetype by lazy {
                            p.duelSettings.queuetype
                                ?: queuetypes.values.first { queuetype -> queuetype?.unranked == true }
                        }
                        val arenas = queuetype?.filterArenasByGametype(it) ?: it.arenas
                        viewer.openMenu(
                            QueueArenaEnableMenu(
                                arenas,
                                { action.execute(viewer) },
                                this
                            )
                        )
                    }, true, this
                )
            )
        }

        setSlot(
            6, ItemStacks.ADVANCED_MODE
        ) { interaction: Interaction ->
            interaction.profile.openMenu(
                MechanicsMenu(
                    this,
                    action
                )
            )
        }
    }

    override val title: String
        get() = CC.BLUE + "Select Mode"

    override fun shouldUpdate() = true
}
